### DISCUSSION TITLE
I tried to extract images  in different formats in a pdf but it extract only 4 images and graph type are   not getting extracted

### DISCUSSION THREAD

**Discussion by kumarlova** [pymupdf/PyMuPDF#2879] [Looking for help]
_**Please provide all mandatory information!**_

## To Reproduce (mandatory)
Explain the steps to reproduce the behavior, For example, include a minimal code snippet, example files, etc.

For problems when building or installing PyMuPDF, give the full output of the build/install command so that, for example, all pip/compiler/linker errors/warnings can be seen.

## Expected behavior (optional)
Describe what you expected to happen (if not obvious).

## Your configuration (mandatory)
 - Operating system, potentially version and bitness
 - Python version, bitness
 - PyMuPDF version, installation method (**wheel** or **generated** from source).

For example, the output of `print(sys.version, "\n", sys.platform, "\n", fitz.__doc__)` would be sufficient (for the first two bullets).

**Comment by kumarlova** [comment]
[Home Gardens in Nepal.pdf](https://github.com/pymupdf/PyMuPDF/files/13631956/Home.Gardens.in.Nepal.pdf)
 this my pdf
 

file = "/content/drive/MyDrive/High voltage data /Home Gardens in Nepal.pdf"
# open the file
pdf_file = fitz.open(file)
 # iterate over pdf pages
for page_index in range(len(pdf_file)):
    # get the page itself
    page = pdf_file[page_index]
    image_list = page.get_images()
    # printing number of images found in this page
    if image_list:
        print(f"[+] Found a total of {len(image_list)} images in page {page_index}")
    else:
        print("[!] No images found on page", page_index)
    for image_index, img in enumerate(page.get_images(), start=1):
        # get the XREF of the image
        xref = img[0]
        # extract the image bytes
        base_image = pdf_file.extract_image(xref)
        image_bytes = base_image["image"]
        # get the image extension
        image_ext = base_image["ext"]
        # load it to PIL
        image = Image.open(io.BytesIO(image_bytes))
        # save it to local disk
        image.save(open(f"image{page_index+1}_{image_index}.{image_ext}", "wb"))
        image

**Comment by JorjMcKie** [comment]
This is no issue, but a Discussions post ... -> converting.

**Reply by JorjMcKie** [reply]
The PDF simply contains only 4 images:
```python
for page in doc:
    print(f"{page}")
    pprint(page.get_images())
    print()

    
page 0 of Home.Gardens.in.Nepal.pdf
[(21, 0, 418, 299, 8, 'DeviceGray', '', 'Im0', 'DCTDecode'),
 (22, 0, 404, 301, 8, 'ICCBased', '', 'Im1', 'DCTDecode')]

page 1 of Home.Gardens.in.Nepal.pdf
[]

page 2 of Home.Gardens.in.Nepal.pdf
[]

page 3 of Home.Gardens.in.Nepal.pdf
[]

page 4 of Home.Gardens.in.Nepal.pdf
[]

page 5 of Home.Gardens.in.Nepal.pdf
[]

page 6 of Home.Gardens.in.Nepal.pdf
[]

page 7 of Home.Gardens.in.Nepal.pdf
[]

page 8 of Home.Gardens.in.Nepal.pdf
[(62, 63, 1066, 12, 8, 'Indexed', '', 'Im0', 'FlateDecode'),
 (64, 0, 985, 549, 8, 'Indexed', '', 'Im1', 'FlateDecode')]

page 9 of Home.Gardens.in.Nepal.pdf
[]

page 10 of Home.Gardens.in.Nepal.pdf
[]

page 11 of Home.Gardens.in.Nepal.pdf
[]

page 12 of Home.Gardens.in.Nepal.pdf
[]
```
Vector graphics cannot be extracted like images, because they are not embedded files or the like. They consist of single draw commands for lines, curves, rectangles, ... with possibly some coloring.
These commands are typically mixed with other display commands of the page to show for example text.
You can try and see if you are able to locate connected clusters of such drawings, join their respective bboxes and then make a picture (Pixmap) of that part of the page - which you then can save as an image file.
There is an example script [here](https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/examples/extract-vector-graphics) that may help you do this.

**Reply by JorjMcKie** [reply]
As a side note:
This utility cannot know what is contained inside the identified rectangles. So it will also extract tables with grid lines, actual images that have an additional border drawn around them, etc.
